cpp/src/interop/ipc.hpp

```diff
@@ -0,0 +1,188 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cinttypes>
+#include <cstring>
+#include <cuda.h>
+#include <cudf/utilities/error.hpp>
+
+namespace cudf {
+namespace ipc {
+inline void check_cu_status(CUresult res)
+{
+  if (res != CUDA_SUCCESS) {
+    char const* msg;
+    if (cuGetErrorString(res, &msg) != CUDA_SUCCESS) {
+      CUDF_FAIL("Unable to get CU error explanation.");
+    }
+    throw logic_error(msg);
+  }
+}
+
+/**
+ * @brief This struct represents a pointer in exported IPC format.
+ */
+struct exported_ptr {
+  cudaIpcMemHandle_t handle{0};
+  int64_t offset{0};
+  int64_t size{0};
+
+  void serialize(std::string* p_bytes) const
+  {
+    std::string& bytes = *p_bytes;
+    size_t this_size   = sizeof(handle) + sizeof(offset) + sizeof(size);
+    size_t orig_size   = p_bytes->size();
+
+    p_bytes->resize(orig_size + this_size);
+    char* ptr = bytes.data() + orig_size;
+
+    std::memcpy(ptr, &handle, sizeof(handle));
+    ptr += sizeof(handle);
+
+    std::memcpy(ptr, &offset, sizeof(offset));
+    ptr += sizeof(offset);
+
+    std::memcpy(ptr, &size, sizeof(size));
+  }
+
+  /**
+   * @brief Construct an `exported_ptr` from buffer returned by `serialize` method.
+   */
+  static uint8_t const* from_buffer(uint8_t const* ptr, exported_ptr* out)
+  {
+    exported_ptr& dptr = *out;
+    std::memcpy(&dptr.handle, ptr, sizeof(handle));
+    ptr += sizeof(handle);
+    std::memcpy(&dptr.offset, ptr, sizeof(offset));
+    ptr += sizeof(offset);
+    std::memcpy(&dptr.size, ptr, sizeof(size));
+    ptr += sizeof(size);
+
+    return ptr;
+  }
+
+  /**
+   * @brief Construct an `exported_ptr` from a given pointer for exportation.
+   */
+  static exported_ptr from_data(uint8_t const* ptr, int64_t size)
+  {
+    /*
+     * `cudaIpcGetMemHandle` returns the same handle for pointers having the same base
+     * address (the address returned from `cudaMalloc`).  So `ptr` and `ptr + 4` have the
+     * same `cudaIpcMemHandle_t`. As a result, we need to calculate the offset from base
+     * address ourseleves and export it along with the handle.
+     */
+    CUdeviceptr pbase;
+    size_t psize;
+    check_cu_status(cuMemGetAddressRange(&pbase, &psize, reinterpret_cast<CUdeviceptr>(ptr)));
+    auto base = reinterpret_cast<uint8_t const*>(pbase);
+
+    cudaIpcMemHandle_t handle;
+    auto non_const = const_cast<void*>(reinterpret_cast<void const*>(ptr));
+    CUDF_CUDA_TRY(cudaIpcGetMemHandle(&handle, non_const));
+
+    return exported_ptr{.handle = handle, .offset = ptr - base, .size = size};
```

Comment:
Designated initializers are not legal C++ before C++20. This will almost certainly generate compiler warnings.

Reply:
Yup, will remove that.

Reply:
Removed.